File: packages/totem-sdk/packages/statechain/src/commitment.ts
import { sha3_256 } from '@totemsdk/core';

export function computeTransferCommitment(
  chainId: string,
  from: string,
  to: string,
  sequence: number,
  timestamp: number,
): Uint8Array {
  const canonical = JSON.stringify({ chainId, from, to, sequence, timestamp });
  return sha3_256(new TextEncoder().encode(canonical));
}

export function computeReclaimCommitment(
  coinId: string,
  ownerPkd: string,
  lockingAddress: string,
): Uint8Array {
  const data = JSON.stringify({ type: 'reclaim', coinId, ownerPkd, lockingAddress });
  return sha3_256(new TextEncoder().encode(data));
}

File: packages/statechain/rust/src/commitment.rs
use sha3::{Digest, Sha3_256};

pub fn compute_transfer_commitment(
    chain_id: &str,
    from: &str,
    to: &str,
    sequence: u32,
    timestamp: u64,
) -> Vec<u8> {
    let canonical = serde_json::json!({
        "chainId": chain_id,
        "from": from,
        "to": to,
        "sequence": sequence,
        "timestamp": timestamp,
    });
    let input = serde_json::to_string(&canonical).unwrap_or_default();
    let mut hasher = Sha3_256::new();
    hasher.update(input.as_bytes());
    hasher.finalize().to_vec()
}

pub fn compute_reclaim_commitment(
    coin_id: &str,
    owner_pkd: &str,
    locking_address: &str,
) -> Vec<u8> {
    let data = serde_json::json!({
        "type": "reclaim",
        "coinId": coin_id,
        "ownerPkd": owner_pkd,
        "lockingAddress": locking_address,
    });
    let input = serde_json::to_string(&data).unwrap_or_default();
    let mut hasher = Sha3_256::new();
    hasher.update(input.as_bytes());
    hasher.finalize().to_vec()
}
